File: shared/core/error-handling/unified-error-handler.ts
/**
 * Unified Error Handling System
 * Consolidates error handling across TypeScript and Python components
 */

import {
  ToolError,
  ToolExecutionResult,
  ToolExecutionStatus,
  ToolContext,
} from '../../types/typescript/tool.types';

export interface ErrorContext {
  toolId: string;
  language: 'typescript' | 'python';
  parameters: any;
  context: ToolContext;
  startTime: Date;
}

export interface ErrorHandler {
  canHandle(error: any): boolean;
  handle(error: any, context: ErrorContext): ToolError;
  priority: number; // Lower number = higher priority
}

export interface RetryStrategy {
  maxAttempts: number;
  backoffMultiplier: number;
  initialDelay: number;
  maxDelay: number;
  shouldRetry: (attempt: number, lastError: ToolError) => boolean;
}

/**
 * Base error handler with common functionality
 */
export abstract class BaseErrorHandler implements ErrorHandler {
  abstract canHandle(error: any): boolean;
  abstract handle(error: any, context: ErrorContext): ToolError;
  abstract priority: number;

  protected createToolError(
    code: string,
    message: string,
    details: any,
    retryable: boolean = false,
  ): ToolError {
    return {
      code,
      message,
      details,
      retryable,
      cause: details?.originalError || details?.original_error,
    };
  }
}

/**
 * Timeout error handler
 */
export class TimeoutErrorHandler extends BaseErrorHandler {
  priority = 1;

  canHandle(error: any): boolean {
    const errorStr = String(error?.message || error).toLowerCase();
    return (
      errorStr.includes('timeout') ||
      error?.code === 'ETIMEDOUT' ||
      error?.code === 'TIMEOUT_ERROR' ||
      (error instanceof Error && error.name === 'TimeoutError')
    );
  }

  handle(error: any, context: ErrorContext): ToolError {
    const duration = Date.now() - context.startTime.getTime();
    return this.createToolError(
      'TOOL_TIMEOUT',
      `Tool execution timed out after ${context.context.timeout}ms`,
      {
        toolId: context.toolId,
        timeout: context.context.timeout,
        duration,
        language: context.language,
      },
      true,
    );
  }
}

/**
 * Validation error handler
 */
export class ValidationErrorHandler extends BaseErrorHandler {
  priority = 2;

  canHandle(error: any): boolean {
    const errorStr = String(error?.message || error).toLowerCase();
    return (
      error?.name === 'ValidationError' ||
      error?.code === 'VALIDATION_ERROR' ||
      errorStr.includes('validation') ||
      errorStr.includes('invalid parameters')
    );
  }

  handle(error: any, context: ErrorContext): ToolError {
    return this.createToolError(
      'VALIDATION_ERROR',
      'Parameter validation failed',
      {
        toolId: context.toolId,
        parameters: context.parameters,
        validationErrors: error.errors || error.details || String(error),
        language: context.language,
      },
      false,
    );
  }
}

/**
 * Permission error handler
 */
export class PermissionErrorHandler extends BaseErrorHandler {
  priority = 3;

  canHandle(error: any): boolean {
    const errorStr = String(error?.message || error).toLowerCase();
    return (
      error?.code === 'EACCES' ||
      error?.code === 'EPERM' ||
      errorStr.includes('permission denied') ||
      errorStr.includes('access denied') ||
      error?.name === 'PermissionError'
    );
  }

  handle(error: any, context: ErrorContext): ToolError {
    return this.createToolError(
      'PERMISSION_DENIED',
      'Permission denied for tool execution',
      {
        toolId: context.toolId,
        operation: context.parameters,
        error: String(error),
        language: context.language,
      },
      false,
    );
  }
}

/**
 * Resource not found error handler
 */
export class ResourceErrorHandler extends BaseErrorHandler {
  priority = 4;

  canHandle(error: any): boolean {
    const errorStr = String(error?.message || error).toLowerCase();
    return (
      error?.code === 'ENOENT' ||
      error?.code === 'ENOTFOUND' ||
      errorStr.includes('not found') ||
      errorStr.includes('does not exist') ||
      error?.name === 'FileNotFoundError'
    );
  }

  handle(error: any, context: ErrorContext): ToolError {
    return this.createToolError(
      'RESOURCE_NOT_FOUND',
      'Required resource not found',
      {
        toolId: context.toolId,
        resource: error.path || error.filename || error.resource || 'unknown',
        error: String(error),
        language: context.language,
      },
      false,
    );
  }
}

/**
 * Language-specific execution error handler
 */
export class ExecutionErrorHandler extends BaseErrorHandler {
  priority = 5;

  canHandle(error: any): boolean {
    return (
      error?.code === 'PYTHON_EXECUTION_ERROR' ||
      error?.code === 'TYPESCRIPT_EXECUTION_ERROR' ||
      error?.source === 'python' ||
      error?.source === 'typescript'
    );
  }

  handle(error: any, context: ErrorContext): ToolError {
    const isRetryable = this.isRetryable(error);
    const code =
      context.language === 'python' ? 'PYTHON_EXECUTION_ERROR' : 'TYPESCRIPT_EXECUTION_ERROR';

    return this.createToolError(
      code,
      String(error?.message || error) || `${context.language} tool execution failed`,
      {
        toolId: context.toolId,
        language: context.language,
        traceback: this.extractTraceback(error),
        stack: error?.stack,
        originalError: String(error),
      },
      isRetryable,
    );
  }

  private extractTraceback(error: any): string[] {
    if (error.traceback) {
      return Array.isArray(error.traceback) ? error.traceback : error.traceback.split('\n');
    }

    const errorString = String(error);
    if (errorString.includes('Traceback')) {
      return errorString.split('\n').filter((line) => line.trim());
    }

    return [];
  }

  private isRetryable(error: any): boolean {
    const errorStr = String(error?.message || error).toLowerCase();

    // Don't retry syntax or import errors
    if (
      errorStr.includes('syntaxerror') ||
      errorStr.includes('importerror') ||
      errorStr.includes('modulenotfounderror') ||
      error instanceof TypeError ||
      error instanceof ReferenceError
    ) {
      return false;
    }

    // Retry network or temporary errors
    if (
      errorStr.includes('connection') ||
      errorStr.includes('timeout') ||
      errorStr.includes('temporary') ||
      error?.code === 'ECONNREFUSED' ||
      error?.code === 'ETIMEDOUT'
    ) {
      return true;
    }

    return false;
  }
}

/**
 * Default fallback error handler
 */
export class DefaultErrorHandler extends BaseErrorHandler {
  priority = 999; // Lowest priority (last resort)

  canHandle(error: any): boolean {
    return true;
  }

  handle(error: any, context: ErrorContext): ToolError {
    return this.createToolError(
      'UNKNOWN_ERROR',
      String(error?.message || error) || 'An unknown error occurred',
      {
        toolId: context.toolId,
        language: context.language,
        error: String(error),
        stack: error?.stack,
        type: error?.constructor?.name || typeof error,
      },
      false,
    );
  }
}

/**
 * Unified error handling middleware
 */
export class UnifiedErrorHandler {
  private handlers: ErrorHandler[] = [];

  constructor() {
    this.registerDefaultHandlers();
  }

  private registerDefaultHandlers(): void {
    this.addHandler(new TimeoutErrorHandler());
    this.addHandler(new ValidationErrorHandler());
    this.addHandler(new PermissionErrorHandler());
    this.addHandler(new ResourceErrorHandler());
    this.addHandler(new ExecutionErrorHandler());
    this.addHandler(new DefaultErrorHandler());
  }

  /**
   * Add a custom error handler
   */
  addHandler(handler: ErrorHandler): void {
    this.handlers.push(handler);
    this.handlers.sort((a, b) => a.priority - b.priority);
  }

  /**
   * Handle an error and convert to ToolExecutionResult
   */
  handleError(error: any, context: ErrorContext): ToolExecutionResult {
    const handler = this.handlers.find((h) => h.canHandle(error));
    if (!handler) {
      throw new Error('No error handler found - this should never happen');
    }

    const toolError = handler.handle(error, context);
    this.logError(toolError, context);

    const endTime = new Date();
    const duration = endTime.getTime() - context.startTime.getTime();

    return {
      toolId: context.toolId,
      executionId: `error_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
      status: ToolExecutionStatus.ERROR,
      error: toolError,
      performance: {
        startTime: context.startTime.toISOString(),
        endTime: endTime.toISOString(),
        duration,
      },
    };
  }

  /**
   * Check if an error is retryable
   */
  isRetryable(error: ToolError): boolean {
    return error.retryable === true;
  }

  /**
   * Create a retry strategy for an error
   */
  createRetryStrategy(error: ToolError): RetryStrategy | null {
    if (!this.isRetryable(error)) {
      return null;
    }

    return {
      maxAttempts: 3,
      backoffMultiplier: 2,
      initialDelay: 1000,
      maxDelay: 30000,
      shouldRetry: (attempt: number, lastError: ToolError) => {
        return attempt < 3 && lastError.retryable === true;
      },
    };
  }

  /**
   * Log error for debugging
   */
  private logError(error: ToolError, context: ErrorContext): void {
    console.error(`[Unified Error Handler] ${error.code}: ${error.message}`, {
      toolId: context.toolId,
      language: context.language,
      retryable: error.retryable,
      details: error.details,
    });
  }
}

// Export singleton instance
export const unifiedErrorHandler = new UnifiedErrorHandler();

// Export factory function for cross-language compatibility
export function createErrorContext(
  toolId: string,
  language: 'typescript' | 'python',
  parameters: any,
  context: ToolContext,
  startTime?: Date,
): ErrorContext {
  return {
    toolId,
    language,
    parameters,
    context,
    startTime: startTime || new Date(),
  };
}

File: shared/core/error-handling/unified_error_handler.py
"""
Unified Error Handling System - Python Implementation
Consolidates error handling across TypeScript and Python components
"""

import traceback
from typing import Any, Dict, Optional, List, Union
from datetime import datetime
from abc import ABC, abstractmethod

from ...types.python.tool import (
    ToolError,
    ToolExecutionResult,
    ToolExecutionStatus,
    ToolContext,
    ToolPerformance
)


class ErrorContext:
    """Context information for error handling"""
    
    def __init__(self, tool_id: str, language: str, parameters: Any, 
                 context: ToolContext, start_time: Optional[datetime] = None):
        self.tool_id = tool_id
        self.language = language
        self.parameters = parameters
        self.context = context
        self.start_time = start_time or datetime.now()


class ErrorHandler(ABC):
    """Base error handler interface"""
    
    @property
    @abstractmethod
    def priority(self) -> int:
        """Handler priority (lower number = higher priority)"""
        pass
    
    @abstractmethod
    def can_handle(self, error: Exception) -> bool:
        """Check if this handler can handle the error"""
        pass
    
    @abstractmethod
    def handle(self, error: Exception, context: ErrorContext) -> ToolError:
        """Handle the error and return a ToolError"""
        pass


class BaseErrorHandler(ErrorHandler):
    """Base error handler with common functionality"""
    
    def create_tool_error(self, code: str, message: str, details: Any, 
                         retryable: bool = False) -> ToolError:
        """Create a ToolError instance"""
        return ToolError(
            code=code,
            message=message,
            details=details,
            retryable=retryable,
            cause=str(details.get('original_error', '')) if isinstance(details, dict) else None
        )


class TimeoutErrorHandler(BaseErrorHandler):
    """Handler for timeout errors"""
    
    @property
    def priority(self) -> int:
        return 1
    
    def can_handle(self, error: Exception) -> bool:
        error_str = str(error).lower()
        return (
            'timeout' in error_str or
            isinstance(error, TimeoutError) or
            (hasattr(error, 'code') and error.code == 'ETIMEDOUT') or
            (hasattr(error, 'code') and error.code == 'TIMEOUT_ERROR')
        )
    
    def handle(self, error: Exception, context: ErrorContext) -> ToolError:
        duration = (datetime.now() - context.start_time).total_seconds() * 1000
        return self.create_tool_error(
            'TOOL_TIMEOUT',
            f'Tool execution timed out after {context.context.timeout}ms',
            {
                'tool_id': context.tool_id,
                'timeout': context.context.timeout,
                'duration': duration,
                'language': context.language
            },
            True
        )


class ValidationErrorHandler(BaseErrorHandler):
    """Handler for validation errors"""
    
    @property
    def priority(self) -> int:
        return 2
    
    def can_handle(self, error: Exception) -> bool:
        error_str = str(error).lower()
        return (
            type(error).__name__ == 'ValidationError' or
            'validation' in error_str or
            'invalid parameters' in error_str or
            (hasattr(error, 'code') and error.code == 'VALIDATION_ERROR')
        )
    
    def handle(self, error: Exception, context: ErrorContext) -> ToolError:
        return self.create_tool_error(
            'VALIDATION_ERROR',
            'Parameter validation failed',
            {
                'tool_id': context.tool_id,
                'parameters': context.parameters,
                'validation_errors': str(error),
                'language': context.language
            },
            False
        )


class PermissionErrorHandler(BaseErrorHandler):
    """Handler for permission errors"""
    
    @property
    def priority(self) -> int:
        return 3
    
    def can_handle(self, error: Exception) -> bool:
        error_str = str(error).lower()
        return (
            isinstance(error, PermissionError) or
            'permission denied' in error_str or
            'access denied' in error_str or
            (hasattr(error, 'code') and error.code in ['EACCES', 'EPERM'])
        )
    
    def handle(self, error: Exception, context: ErrorContext) -> ToolError:
        return self.create_tool_error(
            'PERMISSION_DENIED',
            'Permission denied for tool execution',
            {
                'tool_id': context.tool_id,
                'operation': context.parameters,
                'error': str(error),
                'language': context.language
            },
            False
        )


class ResourceErrorHandler(BaseErrorHandler):
    """Handler for resource not found errors"""
    
    @property
    def priority(self) -> int:
        return 4
    
    def can_handle(self, error: Exception) -> bool:
        error_str = str(error).lower()
        return (
            isinstance(error, FileNotFoundError) or
            'not found' in error_str or
            'does not exist' in error_str or
            (hasattr(error, 'code') and error.code in ['ENOENT', 'ENOTFOUND'])
        )
    
    def handle(self, error: Exception, context: ErrorContext) -> ToolError:
        return self.create_tool_error(
            'RESOURCE_NOT_FOUND',
            'Required resource not found',
            {
                'tool_id': context.tool_id,
                'resource': getattr(error, 'filename', getattr(error, 'path', 'unknown')),
                'error': str(error),
                'language': context.language
            },
            False
        )


class ExecutionErrorHandler(BaseErrorHandler):
    """Handler for language-specific execution errors"""
    
    @property
    def priority(self) -> int:
        return 5
    
    def can_handle(self, error: Exception) -> bool:
        return (
            (hasattr(error, 'code') and error.code in ['PYTHON_EXECUTION_ERROR', 'TYPESCRIPT_EXECUTION_ERROR']) or
            (hasattr(error, 'source') and error.source in ['python', 'typescript'])
        )
    
    def handle(self, error: Exception, context: ErrorContext) -> ToolError:
        is_retryable = self._is_retryable(error)
        code = 'PYTHON_EXECUTION_ERROR' if context.language == 'python' else 'TYPESCRIPT_EXECUTION_ERROR'
        
        return self.create_tool_error(
            code,
            str(error) or f'{context.language} tool execution failed',
            {
                'tool_id': context.tool_id,
                'language': context.language,
                'traceback': self._extract_traceback(error),
                'original_error': str(error)
            },
            is_retryable
        )
    
    def _extract_traceback(self, error: Exception) -> List[str]:
        """Extract traceback information"""
        if hasattr(error, 'traceback'):
            tb = error.traceback
            return tb if isinstance(tb, list) else tb.split('\n')
        
        # Generate traceback for Python exceptions
        if hasattr(error, '__traceback__') and error.__traceback__:
            return traceback.format_exception(type(error), error, error.__traceback__)
        
        return []
    
    def _is_retryable(self, error: Exception) -> bool:
        """Determine if error is retryable"""
        error_str = str(error).lower()
        
        # Don't retry syntax or import errors
        if isinstance(error, (SyntaxError, ImportError, ModuleNotFoundError, TypeError, NameError)):
            return False
        
        if any(err in error_str for err in ['syntaxerror', 'importerror', 'modulenotfounderror']):
            return False
        
        # Retry network or temporary errors
        if any(err in error_str for err in ['connection', 'timeout', 'temporary', 'econnrefused', 'etimedout']):
            return True
        
        return False


class DefaultErrorHandler(BaseErrorHandler):
    """Default fallback error handler"""
    
    @property
    def priority(self) -> int:
        return 999  # Lowest priority
    
    def can_handle(self, error: Exception) -> bool:
        return True
    
    def handle(self, error: Exception, context: ErrorContext) -> ToolError:
        tb = []
        if hasattr(error, '__traceback__') and error.__traceback__:
            tb = traceback.format_exception(type(error), error, error.__traceback__)
        
        return self.create_tool_error(
            'UNKNOWN_ERROR',
            str(error) or 'An unknown error occurred',
            {
                'tool_id': context.tool_id,
                'language': context.language,
                'error': str(error),
                'traceback': tb,
                'type': type(error).__name__
            },
            False
        )


class RetryStrategy:
    """Retry strategy for failed operations"""
    
    def __init__(self, max_attempts: int = 3, backoff_multiplier: float = 2,
                 initial_delay: int = 1000, max_delay: int = 30000):
        self.max_attempts = max_attempts
        self.backoff_multiplier = backoff_multiplier
        self.initial_delay = initial_delay
        self.max_delay = max_delay
    
    def should_retry(self, attempt: int, last_error: ToolError) -> bool:
        """Check if we should retry"""
        return attempt < self.max_attempts and last_error.retryable
    
    def get_delay(self, attempt: int) -> int:
        """Get delay before next retry in milliseconds"""
        delay = self.initial_delay * (self.backoff_multiplier ** attempt)
        return min(int(delay), self.max_delay)


class UnifiedErrorHandler:
    """Unified error handling middleware for tool execution"""
    
    def __init__(self):
        self.handlers: List[ErrorHandler] = []
        self._register_default_handlers()
    
    def _register_default_handlers(self) -> None:
        """Register default error handlers"""
        self.add_handler(TimeoutErrorHandler())
        self.add_handler(ValidationErrorHandler())
        self.add_handler(PermissionErrorHandler())
        self.add_handler(ResourceErrorHandler())
        self.add_handler(ExecutionErrorHandler())
        self.add_handler(DefaultErrorHandler())
    
    def add_handler(self, handler: ErrorHandler) -> None:
        """Add a custom error handler"""
        self.handlers.append(handler)
        self.handlers.sort(key=lambda h: h.priority)
    
    def handle_error(self, error: Exception, context: ErrorContext) -> ToolExecutionResult:
        """Handle an error and convert to ToolExecutionResult"""
        handler = next(h for h in self.handlers if h.can_handle(error))
        tool_error = handler.handle(error, context)
        
        self._log_error(tool_error, context)
        
        end_time = datetime.now()
        duration = int((end_time - context.start_time).total_seconds() * 1000)
        
        return ToolExecutionResult(
            tool_id=context.tool_id,
            execution_id=f"error_{int(datetime.now().timestamp())}_{context.tool_id[:9]}",
            status=ToolExecutionStatus.ERROR,
            error=tool_error,
            performance=ToolPerformance(
                start_time=context.start_time.isoformat(),
                end_time=end_time.isoformat(),
                duration=duration
            )
        )
    
    def is_retryable(self, error: ToolError) -> bool:
        """Check if an error is retryable"""
        return error.retryable
    
    def create_retry_strategy(self, error: ToolError) -> Optional[RetryStrategy]:
        """Create a retry strategy for an error"""
        if not self.is_retryable(error):
            return None
        return RetryStrategy()
    
    def _log_error(self, error: ToolError, context: ErrorContext) -> None:
        """Log error for debugging"""
        print(f"[Unified Error Handler] {error.code}: {error.message}")
        print(f"  Tool ID: {context.tool_id}")
        print(f"  Language: {context.language}")
        print(f"  Retryable: {error.retryable}")
        if error.details:
            print(f"  Details: {error.details}")


# Create singleton instance
unified_error_handler = UnifiedErrorHandler()


def create_error_context(tool_id: str, language: str, parameters: Any,
                        context: ToolContext, start_time: Optional[datetime] = None) -> ErrorContext:
    """Factory function for creating error contexts"""
    return ErrorContext(tool_id, language, parameters, context, start_time)